aiohttp/helpers.py

```diff
@@ -244,6 +243,28 @@ class ProxyInfo:
     proxy_auth: Optional[BasicAuth]
 
 
+def basicauth_from_netrc(netrc_obj: netrc.netrc, host: str) -> Optional[BasicAuth]:
+    auth_from_netrc = netrc_obj.authenticators(host)
+    if auth_from_netrc is None:
+        return None
```

Comment:
It is usually best to have functions return a single type consistently, instead of a variety of types. This saved the caller from having to remember to always call the return value on their side. Instead, exceptions would signal them about the need to process errors explicitly. So I'd do something like this:
```suggestion
def basicauth_from_netrc(netrc_obj: Optional[netrc.netrc], host: str) -> BasicAuth:
    if netrc_obj is None:
        raise LookupError("Unable to locate the `.netrc` file.")
    auth_from_netrc = netrc_obj.authenticators(host)
    if auth_from_netrc is None:
        raise LookupError(f"No entry for {host!s} found in the `.netrc` file.")
```

Reply:
My heuristic for exceptions vs returning none is to get a sense for which is the 'exceptional' circumstance. I would say that *not* finding either a netrc file or an entry for a host is the most common circumstance, even when trust_env is true, while finding an entry is the exceptional circumstance. This is why I prefer it returning None vs exception. There is also a performance penalty here, that is paid for every request when `trust_env` is true. https://switowski.com/blog/ask-for-permission-or-look-before-you-leap/ quantifies this a little, and also suggests the same heuristic I am suggesting - both conceptually and from a performance perspective.

Python 3.11 makes this a bit better with zero cost exceptions (https://bugs.python.org/issue40222) but that only applies if an exception is not actually raised. In this case, as an exception will be raised most of the time so we do not benefit from that optimization.

So my suggestion is to leave this as is. However, given y'all are the maintainers, I will happily fold and implement the suggested changes if you think that is the better way to go :)

Reply:
In the interest of making this easier for you to review, I've implemented your suggestions as a PR to this PR: https://github.com/yuvipanda/aiohttp/pull/2. Merging that will update this PR to use exceptions for control flow. I think it is a better fit, conceptually and from a perf perspective (almost half an ms per request!) to not use exceptions here and leave it as is. But I'm totally happy to be told otherwise, and will merge that PR into this PR :)

Reply:
I'd agree with the performance, but I doubt many users are using `trust_env` and if they enable it, that's probably because there is a good chance of a netrc file being present. So, personally I'm not bothered either way.

Reply:
@Dreamsorcerer i would prefer to keep the `None` if it doesn't bother you each way, as even if they have `trust_env` on, the `netrc` file is unlikely to have an entry for *most* of the hosts requests are being sent to - just some along the way, as it is just Basic Auth. However, I'll wait for @webknjaz's opinion too!

Reply:
I don't think that microoptimizing this is valuable in this case. In Python, EAFP is preferred because it values human readability. This is one of the main motivations of using it in the first place — if you do want to write something high-performance, you'd write it in something like Rust and expose it as a Python C-extension. So for the maintainers, readability/maintainability is often more important than performance, too. The optimization shouldn't be premature and should usually target known/measured bottlenecks. Without any proof that this specific bit is comparable in performance with other activities the framework does, there's no saying that it'll noticably change anything globally. aiohttp has optimizations in places that matter — parsing HTTP traffic, for example.
How often does this happen, anyway? Once per `TCPConnector` creation == once per host? Only when the app sets `trust_env=True` explicitly? Does this mean that if it's set, it's pretty much expected to exist more often than not?
I'd argue that it'd be better to look into caching than microoptimizations. But then, again, it'd be up to the caller/end-user to implement such caching as well as wrap the client object initialization, maybe?

Reply:
@webknjaz I've merged https://github.com/yuvipanda/aiohttp/pull/2 into this PR now, so it uses exceptions for control flow as you had suggested :)

Reply:
I can provide a longer description of the performance impact, but honestly I don't think it particularly matters - the 'is this the exceptional circumstance or expected?' is my usual criteria and performance is just a side benefit. I hope the current changes meet with your approval :)